### PR TITLE
[skstore] Add garbage collection on update

### PR DESCRIPTION
**Pull request by skiplabsdaniel**

### CODE REVIEW

File: skfs/src/Context.sk [hunk 1]
```diff
@@ -1626,4 +1626,31 @@ fun withRegionValue<T>(f: () ~> T): T {
...
+        saved,
+        List[value],
+      ).getHead();
+      !optSaved = Some(SKStore.newObstack());
+    }
+  };
+  SKStore.destroyObstackWithValue(
+    optSaved.fromSome(),
+    List[valueInRegion],
+  ).getHead();
```
Comment (jberdine) [inline skfs/src/Context.sk]: withRegion protects against exceptions, I guess this ought to in a similar way right?

File: compiler/runtime/obstack.c [hunk 2]
```diff
@@ -272,8 +272,18 @@ void SKIP_enable_GC() {
...
+  size_t size = 0;
+  void* cursor = page;
+  while (cursor != NULL && cursor != saved->page) {
+    cursor = *(char**)cursor;
+    nbr_page++;
+    if (nbr_page > 3 || (nbr_page > 1 && (head - page > (2 * PAGE_SIZE / 3)))) {
+      return 1;
+    }
+  }
+  return 0;
```
Comment (jberdine) [inline compiler/runtime/obstack.c:274]: This makes the decision almost not depend on `saved`. Is there any description of the intuition for the new heuristic?

File: skfs/src/Context.sk [hunk 3]
```diff
@@ -1626,4 +1626,31 @@ fun withRegionValue<T>(f: () ~> T): T {
   copied.getHead();
 }
 
+fun withRegionIterate<K, T>(
```
Comment (jberdine) [inline skfs/src/Context.sk]: Since `fn` takes and returns an accumulator, this more of a `fold` than an `iterate`.

File: skfs/src/EagerDir.sk [hunk 4]
```diff
@@ -1576,105 +1577,130 @@ class EagerDir{
...
+    it = dirty.iterator();
+    dKeys = Array<Key>::createFromIterator(it);
+    index = 0;
+    size = dKeys.size();
+    while (index < size) {
+      indexCopy = index;
+      childCopy = childRef;
+      (!index, !childRef) = withRegion(context, (obstack, context) ~> {
+        idx = indexCopy;
+        child = childCopy;
```
Comment (jberdine) [inline skfs/src/EagerDir.sk]: I haven't thought through in detail, but would it be possible to simplify this dance between e.g. `index`, `indexCopy`, and `idx` using your new `withRegionIterate`?

**Review by jberdine — APPROVED** [review]
LGTM, a few minor comments inline.

**Comment by gregsexton** [comment]
I will move the changes in #18 to use withRegionIterate once this lands - this abstraction works well there too! :)

File: skfs/src/Context.sk (hunk 3 above)
Comment (gregsexton) [reply skfs/src/Context.sk]: Ah, I was immediately comfortable here, follows the same naming pattern as e.g. https://clojuredocs.org/clojure.core/iterate

File: skfs/src/Context.sk (hunk 1 above)
Comment (gregsexton) [reply skfs/src/Context.sk]: Yes please. This is my main reason for wanting to replace my manual obstack code in write-csv with something like this.

File: skfs/src/Context.sk (hunk 3 above)
Comment (gregsexton) [reply skfs/src/Context.sk]: I take this back. I misread the fun sig. It's a fold. :)

File: compiler/runtime/obstack.c (hunk 2 above)
Comment (skiplabsdaniel) [reply compiler/runtime/obstack.c:274]: The previous on return always true. It's a try, if you have some intuition on it, you're welcome.

File: skfs/src/Context.sk [hunk 5]
```diff
@@ -1626,4 +1626,52 @@ fun withRegionValue<T>(f: () ~> T): T {
...
+  contextOpt: ?mutable Context,
+  it: mutable Iterator<K>,
+  value: T,
+  fn: (?mutable Context, K, T) ~> T,
+): T {
+  optSaved: ?Obstack = None();
+  valueInRegion = value;
+  loop {
+    if (
+      try {
```
Comment (jberdine) [inline skfs/src/Context.sk]: Am I confused, or could the whole `loop` be in the body of the `try`? I'm not entirely sure of the compilation scheme, but doing that might save some exception setup and cleanup on each loop iteration.

**Review by jberdine — COMMENTED** [review]
Nit: Note misspelling in the commit message.

**Pull request merged**